UnitTests/GitUI.Tests/UserControls/RevisionGrid/Graph/RevisionGraphTests.cs

```diff
@@ -194,5 +392,190 @@ private static IEnumerable<GitRevision> Revisions
                 yield return prevCommit;
             }
         }
+
+        /// <summary>
+        /// Creates a revision graph from a text description that lists commits, space separated, from oldest to newest.
+        /// Each commit spec is {id}:{parent1 id},{parent2 id},...
+        /// or just {id} if it has no parent.
+        ///
+        /// E.g.: " 1   2:1   3:1   4:2,3 "
+        /// </summary>
+        private static RevisionGraph CreateGraph(string commitSpecs)
+        {
+            List<GitRevision> commits = new();
+            Dictionary<string, GitRevision> commitsById = new();
+
+            // Parse and create commits
+            foreach (string spec in commitSpecs.Split(' ', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries))
+            {
+                string[] parts = spec.Split(':');
+                string id = parts[0];
+
+                GitRevision commit = new(ObjectId.Random());
+                commits.Add(commit);
+                commitsById.Add(id, commit);
+
+                if (parts.Length > 1)
+                {
+                    string[] parentIds = parts[1].Split(',');
+                    commit.ParentIds = parentIds.Select(id => commitsById[id].ObjectId).ToList();
+                }
+            }
+
+            // Add commits to graph from newest to oldest
+            RevisionGraph graph = new();
+            commits.Reverse();
+
+            foreach (GitRevision commit in commits)
+            {
+                graph.Add(commit);
+            }
+
+            graph.LoadingCompleted();
+
+            int lastRowIndex = graph.Count - 1;
+            graph.CacheTo(lastRowIndex, lastRowIndex);
+
+            return graph;
+        }
+
+        /// <summary>
+        /// Asserts that the ascii graph of <paramref name="revisionGraph"/> matches <paramref name="expectedLayout"/>.
+        /// For examples of how to specify <paramref name="expectedLayout"/>, see existing usage in tests.
+        /// </summary>
+        private static void AssertGraphLayout(RevisionGraph revisionGraph, string expectedLayout)
+        {
+            string expectedGraph = expectedLayout.Split('\n', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries).Join("\n");
+            string actualGraph = AsciiGraphFor(revisionGraph).Join("\n");
+            try
+            {
+                Assert.AreEqual(expectedGraph, actualGraph);
+            }
+            catch
+            {
+                Console.WriteLine(actualGraph);
+            }
+        }
```

Comment:
Have you considered using snapshot testing instead (i.e., Verify)? This way the diff would be visual

Reply:
The display (of `Console.WriteLine` output) in the test explorer is not perfect, but was good enough for me when using o-kloster's test utility.
Though it is very useful to have the graph definition and the expected result in one place:

```
                RevisionGraph revisionGraph = CreateGraph("R 4:R 3:R 2:R,3 1:2 0:1,4");

                AssertGraphLayout(revisionGraph, @"
                    0
                    |\
                    1 |
                    |  \
                    2   |
                    |\  |
                    | 3 |
                    | | |
                    | | 4
                    |/-´
                    R
                ");
```

Reply:
> The display (of `Console.WriteLine` output) in the test explorer is not perfect, but was good enough for me when using o-kloster's test utility. Though it is very useful to have the graph definition and the expected result in one place:
> 
> ```
>                 RevisionGraph revisionGraph = CreateGraph("R 4:R 3:R 2:R,3 1:2 0:1,4");
> 
>                 AssertGraphLayout(revisionGraph, @"
>                     0
>                     |\
>                     1 |
>                     |  \
>                     2   |
>                     |\  |
>                     | 3 |
>                     | | |
>                     | | 4
>                     |/-´
>                     R
>                 ");
> ```

I am still convinced that this is better tp be kept in one place than fiddling around with lots of `.verified.txt` files.

Reply:
> than fiddling around with lots of `.verified.txt` files.

That's the key point - there's no fiddling required, the verified files are updated automatically. First, if the test result is different from the expectation, you'll see it in a diff tool. Then, if the change is expected, then the new version is applied in the diff tool.